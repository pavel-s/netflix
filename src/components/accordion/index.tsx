import {
  createContext,
  Dispatch,
  SetStateAction,
  useContext,
  useState,
} from 'react';
import { TRSC } from '../../types';
import {
  Container,
  Inner,
  Body,
  Frame,
  Header,
  Item,
  Title,
} from './styles/accordion';

type TAccordion = TRSC<typeof Container> & {
  Title: TRSC<typeof Title>;
  Frame: TRSC<typeof Frame>;
  Item: TRSC<typeof Item>;
  Header: TRSC<typeof Header>;
  Body: TRSC<typeof Body>;
};

const Accordion: TAccordion = ({ children, ...rest }) => {
  return (
    <Container {...rest}>
      <Inner>{children}</Inner>
    </Container>
  );
};

const ToggleContext = createContext<{
  show?: boolean;
  setShow?: Dispatch<SetStateAction<boolean>>;
}>({});

Accordion.Title = ({ children, ...rest }) => (
  <Title {...rest}>{children}</Title>
);

Accordion.Frame = ({ children, ...rest }) => (
  <Frame {...rest}>{children}</Frame>
);

Accordion.Item = function AccordionItem({ children, ...rest }) {
  const [show, setShow] = useState(false);

  return (
    <ToggleContext.Provider value={{ show, setShow }}>
      <Item {...rest}>{children}</Item>
    </ToggleContext.Provider>
  );
};

Accordion.Header = function AccordionHeader({ children, ...rest }) {
  const { show, setShow } = useContext(ToggleContext);

  const toggleShow = () => {
    setShow && setShow((show: boolean) => !show);
  };

  return (
    <Header onClick={toggleShow} {...rest}>
      {children}{' '}
      {show ? (
        <img src='/images/icons/close-slim.png' alt='' />
      ) : (
        <img src='/images/icons/add.png' alt='' />
      )}
    </Header>
  );
};

Accordion.Body = function AccordionBody({ children, ...rest }) {
  const { show } = useContext(ToggleContext);
  return (
    <Body className={show ? 'open' : 'closed'}>
      <span {...rest}>{children}</span>
    </Body>
  );
};

export default Accordion;
