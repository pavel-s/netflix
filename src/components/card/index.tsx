import { TRSC } from '../../types';

import {
  Container,
  CardContainer,
  Group,
  Title,
  Items,
  Item,
  ItemImage,
  Meta,
  Subtitle,
  Text,
  Feature,
  Maturity,
  FeatureClose,
} from './styles/card';

type TCard = TRSC<typeof Container> & {
  Container: TRSC<typeof Container>;
  Group: TRSC<typeof Group>;
  Title: TRSC<typeof Title>;
  Items: TRSC<typeof Items>;
  Item: TRSC<typeof Item>;
  ItemImage: TRSC<typeof ItemImage>;
  Meta: TRSC<typeof Meta>;
  Subtitle: TRSC<typeof Subtitle>;
  Text: TRSC<typeof Text>;
  Feature: TRSC<typeof Feature>;
  Maturity: TRSC<typeof Maturity>;
  FeatureClose: TRSC<typeof FeatureClose>;
};

const Card: TCard = ({ children, ...rest }) => {
  return <CardContainer {...rest}>{children}</CardContainer>;
};

Card.Container = function CardCardContainer({ children, ...rest }) {
  return <Container {...rest}>{children}</Container>;
};

Card.Group = function CardGroup({ children, ...rest }) {
  return <Group {...rest}>{children}</Group>;
};

Card.Title = function CardTitle({ children, ...rest }) {
  return <Title {...rest}>{children}</Title>;
};

Card.Items = function CardItems({ children, ...rest }) {
  return <Items {...rest}>{children}</Items>;
};

Card.Item = function CardItem({ children, ...rest }) {
  return <Item {...rest}>{children}</Item>;
};

Card.ItemImage = function CardItemImage({ children, ...rest }) {
  return <ItemImage {...rest}>{children}</ItemImage>;
};

Card.Meta = function CardMeta({ children, ...rest }) {
  return <Meta {...rest}>{children}</Meta>;
};

Card.Subtitle = function CardSubtitle({ children, ...rest }) {
  return <Subtitle {...rest}>{children}</Subtitle>;
};

Card.Text = function CardText({ children, ...rest }) {
  return <Text {...rest}>{children}</Text>;
};

Card.Feature = function CardFeature({ children, ...rest }) {
  return <Feature {...rest}>{children}</Feature>;
};

Card.Maturity = function CardMaturity({ maturity, ...rest }) {
  return (
    <Maturity maturity={maturity} {...rest}>
      {maturity}
    </Maturity>
  );
};

Card.FeatureClose = function CardFeatureClose({ children, ...rest }) {
  return <FeatureClose {...rest}>{children}</FeatureClose>;
};

export default Card;
