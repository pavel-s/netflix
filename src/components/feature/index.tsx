import { TRSC } from '../../types';
import { Container, Inner, Title, Subtitle } from './styles/feature';

type TFeature = TRSC<'div'> & {
  Title: TRSC<'h1'>;
  Subtitle: TRSC<'h2'>;
};

const Feature: TFeature = ({ children, ...rest }) => {
  return (
    <Container {...rest}>
      <Inner>{children}</Inner>
    </Container>
  );
};

Feature.Title = function FeatureTitle({ children, ...rest }) {
  return <Title {...rest}>{children}</Title>;
};

Feature.Subtitle = function FeatureSubtitle({ children, ...rest }) {
  return <Subtitle {...rest}>{children}</Subtitle>;
};

export default Feature;
