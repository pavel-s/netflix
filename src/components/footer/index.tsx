import { TRSC } from '../../types';
import { Container, Inner, Title, Links, Link, Text } from './styles/footer';

type TFooter = TRSC<'div'> & {
  Title: TRSC<'p'>;
  Links: TRSC<'ul'>;
  Link: TRSC<'a'>;
  Text: TRSC<'p'>;
};

const Footer: TFooter = ({ children, ...rest }) => {
  return (
    <Container {...rest}>
      <Inner>{children}</Inner>
    </Container>
  );
};

Footer.Title = ({ children, ...rest }) => {
  return <Title {...rest}>{children}</Title>;
};

Footer.Links = ({ children, ...rest }) => {
  return <Links {...rest}>{children}</Links>;
};

Footer.Link = ({ children, ...rest }) => {
  return (
    <Link>
      <a {...rest}>{children}</a>
    </Link>
  );
};

Footer.Text = ({ children, ...rest }) => {
  return <Text {...rest}>{children}</Text>;
};

export default Footer;
