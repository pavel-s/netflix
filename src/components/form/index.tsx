import { TRSC } from '../../types';
import {
  Container,
  Base,
  Title,
  Input,
  Error,
  Submit,
  Row,
  Checkbox,
  Text,
  TextBig,
  Link,
  TextButton,
} from './styles/form';

type TForm = TRSC<typeof Container> & {
  Base: TRSC<typeof Base>;
  Title: TRSC<typeof Title>;
  Input: TRSC<typeof Input>;
  Error: TRSC<typeof Error>;
  Submit: TRSC<typeof Submit>;
  Row: TRSC<typeof Row>;
  Checkbox: TRSC<typeof Checkbox>;
  Text: TRSC<typeof Text>;
  TextBig: TRSC<typeof TextBig>;
  Link: TRSC<typeof Link>;
  TextButton: TRSC<typeof TextButton>;
};

const Form: TForm = ({ children, ...rest }) => {
  return <Container {...rest}>{children}</Container>;
};

Form.Base = function FormBase({ children, ...rest }) {
  return <Base {...rest}>{children}</Base>;
};

Form.Title = function FormTitle({ children, ...rest }) {
  return <Title {...rest}>{children}</Title>;
};

Form.Input = function FormInput({ ...rest }) {
  return <Input {...rest} />;
};

Form.Error = function FormError({ ...rest }) {
  return <Error {...rest} />;
};

Form.Submit = function FormSubmit({ isLoading = false, children, ...rest }) {
  return (
    <Submit {...rest} isLoading={isLoading} type='submit'>
      {children}
    </Submit>
  );
};

Form.Row = function FormRow({ children, ...rest }) {
  return <Row {...rest}>{children}</Row>;
};

Form.Checkbox = function FormCheckbox({ children, ...rest }) {
  return <Checkbox {...rest}>{children}</Checkbox>;
};

Form.Text = function FormText({ children, ...rest }) {
  return <Text {...rest}>{children}</Text>;
};

Form.TextBig = function FormTextBig({ children, ...rest }) {
  return <TextBig {...rest}>{children}</TextBig>;
};

Form.Link = function FormLink({ children, ...rest }) {
  return <Link {...rest}>{children}</Link>;
};

Form.TextButton = function FormTextButton({ children, ...rest }) {
  return <TextButton {...rest}>{children}</TextButton>;
};

export default Form;
