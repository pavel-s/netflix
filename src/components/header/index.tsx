import { Background, Container, Logo, SignButton } from './styles/header';
import { LinkProps } from 'react-router-dom';
import { TRSC } from '../../types';

type THeader = TRSC<'div'> & {
  // Container: TRSC<'div'>;
  Logo: TRSC<'img'>;
  Frame: TRSC<'div'>;
  SignButton: TRSC<'a', LinkProps>;
};

const Header: THeader = ({ children, ...rest }) => {
  return <Background {...rest}>{children}</Background>;
};

Header.Frame = function HeaderFrame({ children, ...rest }) {
  return <Container {...rest}>{children}</Container>;
};

Header.Logo = function HeaderLogo({ children, ...rest }) {
  return <Logo {...rest} />;
};

Header.SignButton = function HeaderSignButton({ children, ...rest }) {
  return <SignButton {...rest}>{children}</SignButton>;
};

export default Header;
