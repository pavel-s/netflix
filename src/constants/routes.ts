export const HOME = '/';
export const LOGIN = '/login';
export const LOGIN_HELP = '/LoginHelp';
