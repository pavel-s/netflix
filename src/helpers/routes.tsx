import { FC } from 'react';
import { Redirect, Route, RouteProps } from 'react-router';
import * as ROUTES from '../constants/routes';
import { TFirebaseUser } from '../lib/firebase.prod';

export const IsUserRedirect: FC<
  RouteProps & {
    user: TFirebaseUser | null;
    loggedInPath: typeof ROUTES[keyof typeof ROUTES];
  }
> = ({ user, loggedInPath, children, ...rest }) => {
  return (
    <Route
      {...rest}
      render={() => {
        if (user) {
          return <Redirect to={{ pathname: ROUTES.BROWSE }} />;
        } else {
          return children;
        }
      }}
    />
  );
};

export const ProtectedRoute: FC<RouteProps & { user: TFirebaseUser | null }> =
  ({ user, children, ...rest }) => {
    return (
      <Route
        {...rest}
        render={() => {
          if (user) {
            return children;
          } else {
            return <Redirect to={{ pathname: ROUTES.LOGIN }} />;
          }
        }}
      />
    );
  };
