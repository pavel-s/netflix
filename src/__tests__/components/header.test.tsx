import { fireEvent, render } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import { Header } from '../../components';
import * as ROUTES from './../../constants/routes';

const DEFAULT_USER_PHOTO = 'images/users/1.png';
const BACKGROUND_IMAGE = 'images/misc/joker1.jpg';
const LOGO = 'images/misc/logo.svg';

describe('<Header />', () => {
  it('renders full Header', () => {
    const setCategory = jest.fn();

    const searchTerm = 'frozen';
    const setSearchTerm = jest.fn();

    const profile = {
      displayName: 'Steve',
      photoURL: '/images/users/2.png',
    };

    const signOut = jest.fn();

    const { container, queryByText, queryByAltText, getByText, getByTestId } =
      render(
        <Header url={BACKGROUND_IMAGE} hideOnSmallViewPort>
          <Header.Frame>
            <Header.Group>
              <Header.Logo to={ROUTES.HOME} src={LOGO} alt='Netflix' />
              <Header.TextButton active onClick={() => setCategory('films')}>
                Films
              </Header.TextButton>
              <Header.TextButton onClick={() => setCategory('series')}>
                Series
              </Header.TextButton>
            </Header.Group>
            <Header.Group>
              <Header.Search
                value={searchTerm}
                onChange={(e) => setSearchTerm(e.target.value)}
              />
              <Header.Profile>
                <Header.ProfileImage
                  url={profile.photoURL ? profile.photoURL : DEFAULT_USER_PHOTO}
                />
                <Header.Dropdown>
                  <Header.Group>
                    <Header.ProfileImage
                      url={
                        profile.photoURL ? profile.photoURL : DEFAULT_USER_PHOTO
                      }
                    />
                    <Header.TextButton>{profile.displayName}</Header.TextButton>
                  </Header.Group>
                  <Header.TextButton onClick={signOut}>
                    Sign out
                  </Header.TextButton>
                </Header.Dropdown>
              </Header.Profile>
            </Header.Group>
          </Header.Frame>
          <Header.Feature>
            <Header.FeatureCallOut>Watch Joker now.</Header.FeatureCallOut>
            <Header.Text>Forever alone in a crowd</Header.Text>
            <Header.PlayButton>Play</Header.PlayButton>
          </Header.Feature>
        </Header>,
        { wrapper: MemoryRouter }
      );
    // logo
    expect(queryByAltText('Netflix')).toBeTruthy();

    // category
    const categoryButton = getByText('Films');
    expect(categoryButton).toBeTruthy();
    fireEvent.click(categoryButton);
    expect(setCategory).toBeCalled();

    // search
    const searchInput = getByTestId('search-input');
    expect(searchInput).not.toHaveFocus();
    fireEvent.click(getByTestId('search-button'));
    expect(searchInput).toHaveFocus();

    fireEvent.change(searchInput, {
      target: { value: 'Joker' },
    });
    expect(setSearchTerm).toBeCalledWith('Joker');

    // sign out button
    expect(queryByText('Sign out')).toBeTruthy();

    // feature callout
    expect(queryByText('Watch Joker now.')).toBeTruthy();

    // header text
    expect(queryByText('Forever alone in a crowd')).toBeTruthy();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders basic Header with background', () => {
    const { container, queryByText, queryByAltText } = render(
      <Header url={'/images/misc/home-bg.jpg'}>
        <Header.Frame>
          <Header.Logo src='/images/misc/logo.svg' alt='Netflix' />
          <Header.SignButton to={'/login'}>Sign In</Header.SignButton>
        </Header.Frame>
      </Header>,
      { wrapper: MemoryRouter }
    );

    expect(queryByAltText('Netflix')).toBeTruthy();
    expect(queryByText('Sign In')).toBeTruthy();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders basic Header without background', () => {
    const { container, queryByText, queryByAltText } = render(
      <Header>
        <Header.Frame>
          <Header.Logo src='/images/misc/logo.svg' alt='Netflix' />
          <Header.SignButton to={'/login'}>Sign In</Header.SignButton>
        </Header.Frame>
      </Header>,
      { wrapper: MemoryRouter }
    );

    expect(queryByAltText('Netflix')).toBeTruthy();
    expect(queryByText('Sign In')).toBeTruthy();

    expect(container.firstChild).toMatchSnapshot();
  });
});
