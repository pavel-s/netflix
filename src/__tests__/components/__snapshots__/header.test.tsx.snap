// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> renders basic Header with background 1`] = `
<div
  class="header__Background-sc-1qf3jwc-0 dEhBeJ"
>
  <div
    class="header__Container-sc-1qf3jwc-1 iJrjfW"
  >
    <img
      alt="Netflix"
      class="header__Logo-sc-1qf3jwc-2 edGsUc"
      src="/images/misc/logo.svg"
    />
    <a
      class="header__SignButton-sc-1qf3jwc-4 kCSDSK"
      href="/login"
    >
      Sign In
    </a>
  </div>
</div>
`;

exports[`<Header /> renders basic Header without background 1`] = `
<div
  class="header__Background-sc-1qf3jwc-0 dEAuVS"
>
  <div
    class="header__Container-sc-1qf3jwc-1 iJrjfW"
  >
    <img
      alt="Netflix"
      class="header__Logo-sc-1qf3jwc-2 edGsUc"
      src="/images/misc/logo.svg"
    />
    <a
      class="header__SignButton-sc-1qf3jwc-4 kCSDSK"
      href="/login"
    >
      Sign In
    </a>
  </div>
</div>
`;

exports[`<Header /> renders full Header 1`] = `
<div
  class="header__Background-sc-1qf3jwc-0 ioQOEo"
>
  <div
    class="header__Container-sc-1qf3jwc-1 iJrjfW"
  >
    <div
      class="header__Group-sc-1qf3jwc-9 fRouHy"
    >
      <a
        class="header__LogoLink-sc-1qf3jwc-3 eTWTRb"
        href="/"
      >
        <img
          alt="Netflix"
          class="header__Logo-sc-1qf3jwc-2 edGsUc"
          src="images/misc/logo.svg"
        />
      </a>
      <button
        class="header__TextButton-sc-1qf3jwc-8 fCFpgM"
      >
        Films
      </button>
      <button
        class="header__TextButton-sc-1qf3jwc-8 CpwQL"
      >
        Series
      </button>
    </div>
    <div
      class="header__Group-sc-1qf3jwc-9 fRouHy"
    >
      <div
        class="header__Search-sc-1qf3jwc-15 doTUJA"
      >
        <button
          data-testid="search-button"
        >
          <img
            class="header__SearchIcon-sc-1qf3jwc-13 gFOGps"
            src="images/icons/search.png"
          />
        </button>
        <input
          class="header__SearchInput-sc-1qf3jwc-14 dCGTSp"
          data-testid="search-input"
          type="text"
          value="frozen"
        />
      </div>
      <div
        class="header__Profile-sc-1qf3jwc-12 dJSJdG"
      >
        <div
          class="header__ProfileImage-sc-1qf3jwc-10 bTDlug"
        />
        <div
          class="header__Dropdown-sc-1qf3jwc-11 eGVzkl"
        >
          <div
            class="header__Group-sc-1qf3jwc-9 fRouHy"
          >
            <div
              class="header__ProfileImage-sc-1qf3jwc-10 bTDlug"
            />
            <button
              class="header__TextButton-sc-1qf3jwc-8 CpwQL"
            >
              Steve
            </button>
          </div>
          <button
            class="header__TextButton-sc-1qf3jwc-8 CpwQL"
          >
            Sign out
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="header__Feature-sc-1qf3jwc-5 YbyHX"
  >
    <h2
      class="header__FeatureCallOut-sc-1qf3jwc-6 eUrcDy"
    >
      Watch Joker now.
    </h2>
    <p
      class="header__Text-sc-1qf3jwc-7 eVXLUY"
    >
      Forever alone in a crowd
    </p>
    <button
      class="header__PlayButton-sc-1qf3jwc-16 TubXA"
    >
      Play
    </button>
  </div>
</div>
`;
